api/tokens.go

```diff
@@ -0,0 +1,179 @@
+package api
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/ONSdigital/dp-identity-api/emailvalidation"
+
+	"github.com/ONSdigital/log.go/log"
+)
+
+type ErrorStructure struct {
+	Errors []IndividualError `json:"errors"`
+}
+
+type IndividualError struct {
+	SpecificError string `json:"error"`
+	Message       string `json:"message"`
+	Source        Source `json:"source"`
+}
+
+type Source struct {
+	Field string `json:"field"`
+	Param string `json:"param"`
+}
+
+func TokensHandler() http.HandlerFunc {
+
+	return func(w http.ResponseWriter, req *http.Request) {
+
+		ctx := req.Context()
+
+		field := ""
+		param := ""
+
+		body, err := ioutil.ReadAll(req.Body)
+		if err != nil {
+			errorMessage := "api endpoint POST login returned an error reading the request body"
+			handleUnexpectedError(ctx, w, err, errorMessage, field, param)
+			return
+		}
+
+		defer req.Body.Close()
+
+		authParams := make(map[string]string)
```

Comment:
I think it would be worthwhile creating a type to hold the body data. A map works but it's very open and you could in theory send any JSON data to the endpoint and it would end up being Unmarshalled here.

The other benefit of having a type is that you _could_ (if you wanted to) add validation functions as methods on that type - the data belongs to the model so the model knowing what it is valid feels like reasonable functionality. This is just a suggestion and I'm not specifically requesting this change - I'm just offering food for thought.

However, I do think creating a struct for the request body is necessary.

Reply:
A struct has been created and this is being used rather than a map